### PR TITLE
swiftc: pass -Xcc before -march

### LINKED ISSUE #295322: Build failure: swift for aarch64
Since https://github.com/NixOS/nixpkgs/pull/291901 was merged into the master, swift cannot be built on aarch64.

### Steps To Reproduce

```
nix build github:NixOS/nixpkgs/0dd6aae052b3748d761c93efa16b49c02f8a14e4#swift
```

### Build log

```
Run Build Command(s): /nix/store/40wywa8y2nsgws1lwj7nhmkbbsx1arq4-ninja-1.11.1/bin/ninja -v cmTC_809f1
[1/1] : && /private/tmp/nix-build-swift-5.8.drv-0/build/swift/bin/swiftc -j 8 -num-threads 8 -emit-executable -o cmTC_809f1 -emit-dependencies  -output-file-map CMakeFiles/cmTC_809f1.dir//output-file-map.json  /tmp/nix-build-swift-5.8.drv-0/build/swift-concurrency-backdeploy/CMakeFiles/CMakeScratch/TryCompile-DsXkpX/main.swift    && :
FAILED: cmTC_809f1 CMakeFiles/cmTC_809f1.dir/main.swift.o 
: && /private/tmp/nix-build-swift-5.8.drv-0/build/swift/bin/swiftc -j 8 -num-threads 8 -emit-executable -o cmTC_809f1 -emit-dependencies  -output-file-map CMakeFiles/cmTC_809f1.dir//output-file-map.json  /tmp/nix-build-swift-5.8.drv-0/build/swift-concurrency-backdeploy/CMakeFiles/CMakeScratch/TryCompile-DsXkpX/main.swift    && :
<unknown>:0: error: unknown argument: '-march=armv8.3-a+crypto+sha2+aes+crc+fp16+lse+simd+ras+rdm+rcpc'
ninja: build stopped: subcommand failed.
```
https://hydra.nixos.org/build/252498280
https://hydra.nixos.org/build/252595108

### Additional context
reported in https://github.com/NixOS/nixpkgs/pull/291901#issuecomment-1976557559

I'm not familiar with swift or clang, but I'm using the following overlay as a workaround until it's fixed.
```
self: super: {
  swiftPackages = super.swiftPackages // {
    clang = super.swiftPackages.clang.overrideAttrs (oldAttrs: {
      postFixup =
        (oldAttrs.postFixup or "")
        + ''
          sed -i "s/'-march=.*'//g" $out/nix-support/add-local-cc-cflags-before.sh
        '';
    });
  };
};
 ```

### Notify maintainers
@dtzWill @trepetti @dduan @Trundle @stephank

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.3.0, macOS 14.3.1`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/store/8ny3r3jvj17ljzc89h9rpfr56a7nyn8c-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Samasaur1**
## Description of changes

Adds the `-Xcc` flag before the `-march` flag passed to `swiftc`.

#291901 changed how clang got the `-march` flag, which caused swiftc to get that flag as well. However, swiftc needs `-Xcc` before `-march` because it's a clang flag, not a swift flag.

We could also simply prevent the `-march` flag from being passed to swiftc. I'm not familiar with exactly what this flag does, so I'm choosing to leave it in. However, note that #291901 seems to only have intended to add this flag to clang and accidentally added it to Swift, so maybe we do just want to remove it.

I know adding `-Xcc` made Swift able to build one of my Swift packages on aarch64-darwin, and I assume removing the flag will as well.

Closes #295322.
May supercede #296082.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/compilers/swift/wrapper/wrapper.sh [hunk 1]
```diff
@@ -240,6 +240,11 @@ fi
 # -target, which also happens to work with Swift.
 if [[ -e $cc_wrapper/nix-support/add-local-cc-cflags-before.sh ]]; then
     source $cc_wrapper/nix-support/add-local-cc-cflags-before.sh
+    if [[ "${extraBefore[-1]}" =~ -march=.+ && "${extraBefore[-2]}" != "-Xcc" ]]; then
+        last="${extraBefore[-1]}"
+        extraBefore[-1]="-Xcc"
+        extraBefore+=($last)
+    fi
```
Comment (a-n-n-a-l-e-e) [inline pkgs/development/compilers/swift/wrapper/wrapper.sh:247]: maybe extend the for loop below to handle `-march` anywhere in the array? still fragile if cc-wrapper adds more flags.

```bash
for ((i=0; i < ${#extraBefore[@]}; i++));do
    case "${extraBefore[i]}" in
    -target)
        i=$((i + 1))
        # On Darwin only, need to change 'aarch64' to 'arm64'.
        extraBefore[i]="${extraBefore[i]/aarch64-apple-/arm64-apple-}"
        # On Darwin, Swift requires the triple to be annotated with a version.
        # TODO: Assumes macOS.
        extraBefore[i]="${extraBefore[i]/-apple-darwin/-apple-macosx${MACOSX_DEPLOYMENT_TARGET:-11.0}}"
        ;;
    -march=*)
        [[ i -gt 0 && ${extraBefore[i-1]} == -Xcc ]] && continue
        extraBefore=(
            "${extraBefore[@]:0:i}"
            -Xcc
            "${extraBefore[@]:i:${#extraBefore[@]}}"
        )
        i=$((i + 1))
        ;;
    esac
done
```

Comment (Samasaur1) [reply pkgs/development/compilers/swift/wrapper/wrapper.sh:247]: Ah, that's much nicer. Let me test this and if it works, I'll amend and force-push

**Comment by alyssais** [comment]
Haven't reviewed in depth because I don't know about swift, but I like the approach.  -march tells the compiler to optimise for a specific architecture, so we probably want it for Swift.

**Comment by a-n-n-a-l-e-e** [comment]
@ofborg build swift

**Review by ghost — APPROVED** [review]

**Comment by a-n-n-a-l-e-e** [comment]
x64/aarch64 darwin builds pass on community builder

**Pull request merged**